### PR TITLE
Export keytype and valtype for Associative Collections 

### PR DESCRIPTION
**Pull request by mdcfrancis**
0.4 changes dict from tuple to Pair{keytype,valtype} this causes access to the key and value type to be obscure. There already exist functions which are not currently exported. This pull requests exports these

https://groups.google.com/forum/#!topic/julia-users/s1c7crYgctY

Pull request includes doc updates, including a few existing errors and also documents the new functions. Tests have been added to ensure that the newly exported functions work correctly.

### CODE REVIEW

File: base/docs/helpdb.jl [hunk 1]
```diff
@@ -1466,11 +1466,25 @@ range
 doc"""
     eltype(type)
 
-Determine the type of the elements generated by iterating a collection of the given `type`. For associative collection types, this will be a `(key,value)` tuple type. The definition `eltype(x) = eltype(typeof(x))` is provided for convenience so that instances can be passed instead of types. However the form that accepts a type argument should be defined for new types.
+Determine the type of the elements generated by iterating a collection of the given `type`. For associative collection types, this will be a `Pair{keytype,valtype}`. The definition `eltype(x) = eltype(typeof(x))` is provided for convenience so that instances can be passed instead of types. However the form that accepts a type argument should be defined for new types.
```
Comment (phobon) [inline base/docs/helpdb.jl]: You should change the doc references of `keytype` and `valtype` so they aren't ambiguous with the function names. Probably some capitalization would do it.

Comment (mdcfrancis) [inline base/docs/helpdb.jl]: I have updated as suggested

**Comment by mdcfrancis** [comment]
Updates to docs to capitalize types in the examples.

**Comment by tkelman** [comment]
Best to squash this and rebase out the merge commits

**Comment by mdcfrancis** [comment]
@tkelman can you point me in the right direction to do that ?

**Comment by tkelman** [comment]
Assuming you have this branch checked out, a remote called `JuliaLang` for this upstream repository, and a remote called `mdcfrancis` for your fork, it should look like

```
git rebase -i JuliaLang/master
# change pick to squash for all but one, maybe two (docs separate from code changes perhaps?) of the commits - you can also reorder commits here
# check git log, diff of last few commits to make sure everything looks right
git push -f mdcfrancis mdf_dict_exports
```

Then refresh the page here to check that it went through as you intended. If it doesn't, we can always get back to the existing commit (441134a5d00bdb4a1edb090c4d44c84428e7c3e1) so it's not the end of the world if it doesn't go through correctly the first time.

**Comment by mdcfrancis** [comment]
@tkelman git it thanks - I have rebased and updated.

**Pull request merged**